### PR TITLE
Add shapely stubs

### PR DESCRIPTION
**Pull request by hamdanal**
I basically copied the stubs from https://github.com/hamdanal/python-stubs/tree/5ecdeaede3a5d97b00b5d4298f13b7028c01e6eb/stubs/shapely-stubs

Note that these stubs are somewhat complete as I've been working on them for quiet some time. There is an extensive [test suite](https://github.com/hamdanal/python-stubs/tree/5ecdeaede3a5d97b00b5d4298f13b7028c01e6eb/tests/shapely) which passes both mypy and pyright in addition to runtime type checking.  I've also been testing those stubs in VS Code (through Pylance, no actual type checking) for a while and they seem to work as expected. Hopefully this will make the review a bit easier.

I haven't ported any of the tests to typeshed for now. I might add a subset of them later to avoid regressions.

Technical notes:
* The stubs contain a lot of overloads. This is because shapely implements the numpy universal functions (ufunc) API and the behavior of its functions depend on whether they are passed scalar values as arguments or array-like objects.
* The type hierarchy of objects in shapely is as follows: "All shapely objects" -> `BaseGeometry` -> `Geometry` -> `object`. For input types, I use `shapely.Geometry` as much as possible. For output types, I use  `shapely.geometry.BaseGeometry`. This is because `BaseGeometry` is an implementation detail that should not be required to be imported by users -- they should be able to annotate their code using `Geometry`. For output types, `Geometry` is not very useful as it doesn't have any attributes or methods.

Related: https://github.com/shapely/shapely/pull/1741#issuecomment-1767298521

### CODE REVIEW

File: stubs/shapely/shapely/__init__.pyi [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    geos_version as geos_version,
+    geos_version_string as geos_version_string,
+)
+from .linear import *
+from .measurement import *
+from .predicates import *
+from .set_operations import *
+from .strtree import *
+
+__version__: str
```
Comment (srittau) [inline stubs/shapely/shapely/__init__.pyi:32]: Nit (needs import):

```suggestion
__version__: Literal[str]
```

File: stubs/shapely/shapely/_typing.pyi [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    @property
+    def __geo_interface__(self) -> dict[str, Any]: ...
+
+@type_check_only
+class SupportsRead(Protocol[_T_co]):
+    def read(self) -> _T_co: ...
+
+@type_check_only
+class SupportsWrite(Protocol[_T_contra]):
+    def write(self, s: _T_contra, /) -> object: ...
```
Comment (srittau) [inline stubs/shapely/shapely/_typing.pyi:61]: Considering that these are copies from `_typeshed`, maybe we should just re-export them? (Not that it matters terribly.)

**Review by srittau — APPROVED** [review]
Glancing through this, I see no red flags. Two optional suggestions below.

File: stubs/shapely/shapely/__init__.pyi (hunk 1 above)
Comment (hamdanal) [reply stubs/shapely/shapely/__init__.pyi]: I think you meant `Final`. Fixed.

File: stubs/shapely/shapely/_typing.pyi (hunk 2 above)
Comment (hamdanal) [reply stubs/shapely/shapely/_typing.pyi]: I deleted these definitions and used the typeshed one's instead. There is no need to redefine them.

**Review by srittau — APPROVED** [review]

**Pull request merged**